### PR TITLE
Integration with ODD

### PR DESCRIPTION
**Pull request by iliax**
- [ ] **Breaking change?** (if so, please describe the impact and migration path for existing application instances)

Exporting to ODD:
- topics as DataSets (metadata + schemas)
- connectors as DataTransformers (metadata + inputs&outputs)

**What changes did you make?** (Give an overview)

**Is there anything you'd like reviewers to focus on?**

**How Has This Been Tested?** (put an "x" (case-sensitive!) next to an item)

- [ ] No need to
- [x] Manually (please, describe, if necessary)
- [x] Unit checks
- [ ] Integration checks
- [ ] Covered by existing automation

**Checklist** (put an "x" (case-sensitive!) next to all the items, otherwise the build will fail)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (e.g. **ENVIRONMENT VARIABLES**)
- [x] My changes generate no new warnings (e.g. Sonar is happy)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged

Check out [Contributing](https://github.com/provectus/kafka-ui/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/provectus/kafka-ui/blob/master/CODE-OF-CONDUCT.md)

**A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/ConnectorsExporter.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    return new DataEntity()
+        .oddrn(Oddrn.connectorOddrn(connect.getAddress(), connector.getName()))
+        //TODO[discuss]: name generation (maybe include connect/cluster name)
+        .name("Kafka Connector \"%s\" (%s)".formatted(connector.getName(), connector.getType()))
+        .type(DataEntityType.JOB)
+        .dataTransformer(transformer)
+        .metadata(List.of(
+            new MetadataExtension()
+                .schemaUrl(URI.create("wontbeused.oops"))
+                .metadata(info.metadata())));
```
Comment (iliax) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/ConnectorsExporter.java:46]: - add all configs to metadata (with sanitizing)
- check topics list endpoint in connector

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/OddExporter.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    String clusterOddrn = Oddrn.clusterOddrn(cluster);
+    return oddApi.getDataEntitiesByDEGOddrn(clusterOddrn)
+        .map(r -> !r.getItems().isEmpty())
+        .onErrorResume(WebClientResponseException.NotFound.class, throwable -> Mono.just(false))
+        .filter(created -> !created)
+        .flatMap(notFound ->
+            oddApi.createDataSource(
+                new DataSourceList()
+                    .addItemsItem(
+                        new DataSource()
```
Comment (iliax) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/OddExporter.java:98]: create DS on startup without existence checks

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/OddExporter.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+                            .oddrn(clusterOddrn)
+                            .name("Kafka cluster \"%s\"".formatted(cluster.getName()))
```
Comment (iliax) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/OddExporter.java:100]: cluster.getName()

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/TopicsExporter.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        .flatMap(topic -> createTopicDataEntity(cluster, topic, stats));
+  }
+
+  @SneakyThrows
+  private Mono<DataEntity> createTopicDataEntity(KafkaCluster cluster, String topic, Statistics stats) {
+    KafkaPath topicOddrnPath = Oddrn.topicOddrnPath(cluster, topic);
+    return getTopicSchema(cluster, topic, topicOddrnPath, false)
+        .map(valueFields ->
+            new DataEntity()
+                .name("Topic \"%s\"".formatted(topic)) //TODO[discuss]: discuss naming
```
Comment (iliax) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/TopicsExporter.java:49]: raw topic name

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/Oddrn.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            .build(),
+        "key"
+    );
+  }
+
+  String connectorOddrn(String connectUrl, String connectorName) {
+    String connectorHost = Stream.of(connectUrl.split(","))
+        .map(String::trim)
+        .sorted()
+        //TODO[discuss]: leaving host and [port] in oddrn
```
Comment (iliax) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/Oddrn.java:55]: good

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/TopicsExporter.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  private Map<String, Object> getTopicMetadata(String topic, Statistics stats) {
+    TopicDescription topicDescription = stats.getTopicDescriptions().get(topic);
+    return ImmutableMap.<String, Object>builder()
+        .put("partitions", topicDescription.partitions().size())
+        .put("replication_factor", topicDescription.partitions().get(0).replicas().size())
+        .putAll(getNonDefaultConfigs(topic, stats))
+        .build();
+  }
+
+  private Mono<List<DataSetField>> getTopicSchema(KafkaCluster cluster,
```
Comment (iliax) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/TopicsExporter.java:81]: also export keys schema

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/schema/AvroExtractor.java [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+  private static void extractUnion(Schema schema,
+                                   String parentOddr,
+                                   String oddrn, //null for root
+                                   String name,
+                                   String doc,
+                                   ImmutableSet<String> registeredRecords,
+                                   List<DataSetField> sink) {
+    boolean isRoot = oddrn == null;
+    boolean containsNull = schema.getTypes().stream().map(Schema::getType).anyMatch(t -> t == Schema.Type.NULL);
+    //TODO[discuss]: how this approach will work with backw-compat check? ->
```
Comment (iliax) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/schema/AvroExtractor.java:111]: works fine

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/schema/ProtoExtractor.java [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+  private static void extractRepeated(Descriptors.FieldDescriptor field,
+                                      String parentOddr,
+                                      String oddrn, //null for root
+                                      String name,
+                                      boolean nullable,
+                                      ImmutableSet<String> registeredRecords,
+                                      List<DataSetField> sink) {
+    sink.add(createDataSetField(name, parentOddr, oddrn, TypeEnum.LIST, "repeated", nullable));
+
+    String itemName = field.getType() == Descriptors.FieldDescriptor.Type.MESSAGE
```
Comment (iliax) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/schema/ProtoExtractor.java:125]: extract to method

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/schema/ProtoExtractor.java [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+      sink.add(createDataSetField(name, parentOddr, oddrn, TypeEnum.DATETIME, typeName, nullable));
+      return true;
+    }
+    if (typeName.equals(Duration.getDescriptor().getFullName())) {
+      sink.add(createDataSetField(name, parentOddr, oddrn, TypeEnum.DURATION, typeName, nullable));
+      return true;
+    }
+    if (typeName.equals(Value.getDescriptor().getFullName())) {
+      //TODO[discuss] : mapping Value to UNION type (maybe its better to use UNKNOWN?)
+      sink.add(createDataSetField(name, parentOddr, oddrn, TypeEnum.UNION, typeName, nullable));
```
Comment (iliax) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/schema/ProtoExtractor.java:105]: UNKNOW, future(ANY)

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/schema/JsonSchemaExtractor.java [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+    if (schema.getCriterion() == CombinedSchema.ONE_CRITERION) {
+      combineType = "oneOf";
+    }
+
+    boolean isRoot = oddrn == null;
+    oddrn = isRoot ? (parentOddr + "/" + combineType) : (oddrn + "/" + combineType);
+    sink.add(
+        createDataSetField(
+            schema,
+            isRoot ? "Root %s".formatted(combineType) : "%s (%s)".formatted(name, combineType),
```
Comment (iliax) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/service/integration/odd/schema/JsonSchemaExtractor.java:230]: rm combineType from name to field metadata

**Review by Haarolean — APPROVED** [review]

**Pull request merged**